R/cube-dims.R

```diff
@@ -144,29 +144,84 @@ setMethod("[", "CubeDims", function (x, i, ...) {
-            x$type == "categorical" &&
-            length(dim$name) == 3
-        if (maybe) {
-            cats <- Categories(data=x$categories)
-            ## Unlike the strict is.3vl, this doesn't compare cat names because
-            ## they've already been munged to TRUE/FALSE
-            maybe <- setequal(ids(cats), c(-1, 0, 1)) &&
-                sum(is.selected(cats)) == 1 &&
-                sum(is.na(cats)) == 1
+is.selectedDimension <- function (dims) getDimTypes(dims) == "mr_selections"
+
+#' Get dimension type
+#'
+#' This function returns the specific type of each cube dimension. This is useful
+#' when cubes contain categorical array or multiple response variables because it
+#' identifies the dimensions of the cube which refer to the different parts of
+#' array variable:
+#' - `ca_items`: Categorical array items
+#' - `ca_categories`: The categories of the categorical array
+#' - `mr_items`: Multiple response options or items
+#' - `mr_selections`: The selection status for a multiple response variable
+#'
+#' @return A character vector. This is identical to `types()` except that
+#' the array variable types are more specific.
+#' @param x a CrunchCube or CubeDims object
+#'
+#' @return a character vector of dimension types
+#' @export
+#' @keywords internal
+getDimTypes <-  function (x) {
+    if (inherits(x, "CrunchCube")) {
+        x <- x@dims
+    }
+
+    what_dim_is_it <- function (one_var, array_aliases) {
+        dim_type <- type(one_var)
+        
+        if (alias(one_var) %in% array_aliases) {
+            # we are in an array, we need to figure out if this is a multiple
+            # response or not
+            array_cat_dim <- vars[aliases(vars) == alias(one_var) & types(vars) == "categorical"]
+
+            # if we can't find the matching categories dimension we might have a
+            # subset cube, so simply return the dim_type un-identified (this
+            # might could actually just be "ca_items")
+            if (length(array_cat_dim) < 1) {
+                return(dim_type)
+            }
+            
+            # if this is a variable crossed by itself, then array_cat_dim will
+            # actually have two copies of the categories dimension. We take the
+            # first one becasue it should be identical to all the others. If it
+            # isn't this might produce weird results. 
+            # TODO: Investigate checking by ID
+            array_cats <- categories(array_cat_dim[[1]])
+            
+            # if we meet these conditions, we are actually a multiple response
+            # and should label ourself as such.
+            # Unlike the strict is.3vl, this doesn't compare cat names because
+            # they've already been munged to TRUE/FALSE
+            is.MR <- length(array_cats) == 3 &&
+                setequal(ids(array_cats), c(-1, 0, 1)) &&
+                sum(is.selected(array_cats)) == 1 &&
+                sum(is.na(array_cats)) == 1
```

Comment:
I believe `is.MR <- sum(is.selected(array_cats)) >= 1` should be sufficient.

Reply:
I _think_ if we did that, we wouldn't be able to query the MR as a categorical array, ignoring the "selected" bit.

Reply:
Yeah, I shared the same suspicion, but haven't had an opportunity to investigate if it's true. At the very least, the (possibly overly)-specific version doesn't stop us from identifying correctly MRs in place now, so I'm inclined to keep it, at least for now.

Reply:
You would, you'd just have to have wrap it in `select_categories(mr, [])` which would strip any "selected" flags. I believe that's a rare need (correct me if I'm wrong) so it seems better to me to wrap with an extra function for that than to make every other call more complex.

Reply:
We can (in a separate PR) switch to using `select_categories(mr, [])`, and then in principle we could simplify this check (the computational cost of which is not a problem), but we'd also have to teach the cube logic that there could be more than one "selected" category and aggregate accordingly. Definitely out of scope here though.